Introduce custom sympy srepr parser [stable/2.0]

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit introduces a custom parser to QPY for parameter expression payloads that were generated using sympy. Prior to QPY version 10 this was the only way we supported serializing parameter expressions in QPY. For QPY version 10, 11, and 12 sympy could optionally be used if the payload was generated explicitly to not use symengine (in qiskit 1.0 it defaulted to use symengine).

This serialization format relied on sympy to generate a string representation of the expression which we then put in the payload. On deserialization we called sympy's parse_expr() function which internally is calling sympy's sympify() internally. Sympy documents that sympify() relies on Python's eval() for string input and should not be used with untrusted input. But parse_expr() didn't have such a warning (at the time, I plan to contribute adding one), so using this function provides an avenue for arbitrary code execution during QPY deserialization.

This commit fixes this issue by writing a custom parser for the string representation in a QPY payload based on python's ast module. This parser walks the abstract syntax tree and builds the sympy expression object as it it goes. It is restricted to the operations that ParameterExpression supports and if any part of the string tries to use functionality outside that set it will error.

### Details and comments

mtreinish committed:
Introduce custom sympy srepr parser

This commit introduces a custom parser to QPY for parameter expression
payloads that were generated using sympy. Prior to QPY version 10 this
was the only way we supported serializing parameter expressions in QPY.
For QPY version 10, 11, and 12 sympy could optionally be used if the
payload was generated explicitly to not use symengine (in qiskit 1.0 it
defaulted to use symengine).
This serialization format relied on sympy to generate a string
representation of the expression which we then put in the payload. On
deserialization we called sympy's `parse_expr()` function which
internally is calling sympy's `sympify()` internally. Sympy documents
that `sympify()` relies on Python's `eval()` for string input and
should not be used with untrusted input. But `parse_expr()` didn't have
such a warning (at the time, I plan to contribute adding one), so
using this function provides an avenue for arbitrary code execution
during QPY deserialization.

This commit fixes this issue by writing a custom parser for the string
repesentation in a QPY payload based on python's ast module. This parser
walks the abstract syntax tree and builds the sympy expression object as
it it goes. It is restricted to the operations that
`ParameterExpression` supports and if any part of the string tries to
use functionality outside that set it will error.

mtreinish committed:
Simplify visitor logic

mtreinish committed:
Fix lint

mtreinish committed:
Add sanity checks that we only call sympify if coming from symengine

ElePT commented on test/python/qpy/test_circuit_load_from_qpy.py:
```diff
@@ -289,6 +289,7 @@ def test_rdiv(self):
         self.assert_roundtrip_equal(qc)
 
 
+@ddt
```
```suggestion
```
oversight?

mtreinish replied:
Yeah this was a leftover from the 1.4.2 PR's testing which no longer applies to the 2.0.0 branch because we've dropped exporting qpy < 13.

ElePT commented on releasenotes/notes/fix-ace-qpyv1-609551d84531e016.yaml:
```diff
@@ -0,0 +1,9 @@
+    :class:`.ParameterExpression` objects and other symbolic expressions.
+    This potentially includes any QPY payload using QPY version < 10, and
+    optionally 10, 11, and 12 depending on the symbolic encoding used in the
+    payload.
```
Same suggestion as in the 1.4.2 reno:

```suggestion
    Fixed a security vulnerability in :func:`.qpy.load` when loading
    payloads that use ``sympy`` to serialize
    :class:`.ParameterExpression` objects and other symbolic expressions.
    This potentially includes any QPY payload using QPY version < 10, and
    optionally 10, 11, and 12 depending on the symbolic encoding used in the
    serialization step (:func:`.qpy.dump`).
```

ElePT reviewed (commented):
LGTM! I just have two minor comments and then I think it's ready.

ElePT commented:
Is the plan to also add version changes to this PR and use it to release like in #14022?

mtreinish commented:
> Is the plan to also add version changes to this PR and use it to release like in #14022?

Let me do that now and apply your suggestions, that will save us some CI time.

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

mtreinish committed:
Bump version numbers to prepare rc2

ElePT reviewed (dismissed):
LGTM

Cryoris committed:
Fix PGO build

PGO fails when we build qiskit-cext both as cdylib in addition to rlib,
due to missing Python symbols. This could be due to us not setting the
pyo3/extension-module correctly; which we want enabled when we compile
the crate for use in the Python extension crate qiskit-pyext, but
don't want for the standalone cdylib usage.

This PR changes to only building the rlib per default, meaning that
building the qiskit-pyext crate (and running PGO) will not build the
cdylib. To build the cdylib for C standalone mode, one has to request
that crate type via cargo rustc .... --crate-type cdylib. We're
investigating further solutions but this now fixes PGO and the C
standalone tests pass as well.
